crates/sail-execution/src/rpc.rs

```diff
@@ -119,35 +119,41 @@ impl_client_builder!(DriverServiceClient<Channel>);
 impl_client_builder!(WorkerServiceClient<Channel>);
 impl_client_builder!(FlightServiceClient<Channel>);
 
+/// A handle to a gRPC client to support connection reuse.
+/// The handle can be cheaply cloned and the underlying connection is shared.
 #[derive(Debug, Clone)]
 pub struct ClientHandle<T> {
+    /// The client options.
     options: Arc<ClientOptions>,
-    inner: OnceCell<Arc<Mutex<T>>>,
+    /// The shared gRPC client which is lazily initialized.
+    /// Note that this must be `Arc<OnceCell<T>>` instead of `OnceCell<Arc<T>>`.
+    /// If we use the latter, when the client is not initialized, an empty `OnceCell` would be
+    /// cloned and later initialized independently, resulting in multiple connections.
+    /// This could then easily overwhelm the server, and the client would see the
+    /// "connection refused" Tonic transport error.
```

Comment:
This was the subtle bug that caused "transport error" when cluster load is high.